common/src/main/java/org/opensearch/ml/common/model/MLGuard.java

```diff
@@ -0,0 +1,133 @@
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.builder.SearchSourceBuilder;
+
+import java.security.AccessController;
+import java.security.PrivilegedExceptionAction;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static org.opensearch.ml.common.utils.StringUtils.gson;
+
+@Log4j2
+@Getter
+public class MLGuard {
+    private Map<String, List<String>> stopWordsIndicesInput = new HashMap<>();
+    private Map<String, List<String>> stopWordsIndicesOutput = new HashMap<>();
+    private List<String> inputRegex;
+    private List<String> outputRegex;
+    private NamedXContentRegistry xContentRegistry;
+    private Client client;
+
+    public MLGuard(Guardrails guardrails, NamedXContentRegistry xContentRegistry, Client client) {
+        this.xContentRegistry = xContentRegistry;
+        this.client = client;
+        if (guardrails == null) {
+            return;
+        }
+        Guardrail inputGuardrail = guardrails.getInputGuardrail();
+        Guardrail outputGuardrail = guardrails.getOutputGuardrail();
+        if (inputGuardrail != null) {
+            fillStopWordsToMap(inputGuardrail, stopWordsIndicesInput);
+            inputRegex = inputGuardrail.getRegex() == null ? new ArrayList<>() : Arrays.asList(inputGuardrail.getRegex());
+        }
+        if (outputGuardrail != null) {
+            fillStopWordsToMap(outputGuardrail, stopWordsIndicesOutput);
+            outputRegex = outputGuardrail.getRegex() == null ? new ArrayList<>() : Arrays.asList(outputGuardrail.getRegex());
+        }
+    }
+
+    private void fillStopWordsToMap(@NonNull Guardrail guardrail, Map<String, List<String>> map) {
+        List<StopWords> stopWords = guardrail.getStopWords();
+        if (stopWords == null || stopWords.isEmpty()) {
+            return;
+        }
+        for (StopWords e : stopWords) {
+            map.put(e.getIndex(), Arrays.asList(e.getSourceFields()));
+        }
+    }
+
+    public Boolean validate(String input, int type) {
+        switch (type) {
+            case 0: // validate input
+                return validateRegexList(input, inputRegex) && validateStopWords(input, stopWordsIndicesInput);
+            case 1: // validate output
+                return validateRegexList(input, outputRegex) && validateStopWords(input, stopWordsIndicesOutput);
+            default:
+                return true;
+        }
+    }
+
+    public Boolean validateRegexList(String input, List<String> regexList) {
+        for (String regex : regexList) {
+            if (!validateRegex(input, regex)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public Boolean validateRegex(String input, String regex) {
+        Pattern pattern = Pattern.compile(regex);
```

Comment:
Improve the efficiency of this code ? We can compile the regex pattern only once rather than every time you validate the text. The regex compiling operation can be costly, and unnecessary compiling should be avoided.

Reply:
make sense

Reply:
Stored the list of pattern in the mlguard object to reuse.